index.js

```diff
@@ -0,0 +1,555 @@
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.amount - An amount.
+ *
+ * @throws DataError - If the currency is not supported by PayPal.
+ * @throws DataError - If the amount's value is not a number.
+ *
+ * @returns {object} A formatted amount.
+ */
+const formatAmount = ({amount}) => {
+  const supported = currencies.supported.has(amount.currency_code);
+  if(!supported) {
+    throw new BedrockError(
+      `Unsupported PayPal currency ${amount.currency_code}.`,
+      Errors.Data, {public: true}
+    );
+  }
+  const bigAmount = BigNumber(amount.value);
+  const notNumber = bigAmount.toString() === 'NaN';
+  if(notNumber) {
+    throw new BedrockError(
+      `Invalid amount ${amount.value} ${amount.currency_code}.`,
+      Errors.Data, {public: true}
+    );
+  }
+  if(currencies.noDecimal.has(amount.currency_code)) {
+    amount.value = bigAmount.toFixed(0).toString();
+    return amount;
+  }
+  amount.value = bigAmount.toFixed(2).toString();
+  return amount;
+};
+
+/**
+ * Gets a PayPal auth token so we can call on the api.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} [options.clientId = paypal_client_id] - PayPal Client Id.
+ * @param {string} [options.secret = paypal_secret] - PayPal Secret.
+ *
+ * @returns {Promise<object>} The data returned with an `access_token`.
+ */
+const getAuthToken = async ({clientId, secret, api}) => {
+  const authDataKey = 'authData';
+  const authCache = paypalCache.get(authDataKey);
+  if(authCache) {
+    return authCache;
+  }
+  const authUrl = `${api}/v1/oauth2/token`;
+  const options = {
+    headers: {
+      Accept: 'application/json',
+      'Content-Type': 'application/x-www-form-urlencoded',
+      'Accept-Language': 'en_US'
+    },
+    auth: {
+      username: clientId,
+      password: secret
+    }
+  };
+  const body = 'grant_type=client_credentials';
+  try {
+    const {data} = await axios.post(authUrl, body, options);
+    // cache expires when PayPal says it will expire.
+    const {expires_in} = data;
+    // expire the cache 5 seconds before PayPal's timeout.
+    // or just don't have a time out.
+    const expires = expires_in > 5 ? expires_in - 5 : 0;
+    paypalCache.set(authDataKey, data, expires);
+    return data;
+  } catch(error) {
+    const {errorType, details} = formatAxiosError({error});
+    throw new BedrockError('PayPal Authentication failed.', errorType, details);
+  }
+};
+
+/**
+ * Gets the options needed to make a secure paypal request.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} [options.clientId = paypal_client_id] - PayPal Client Id.
+ * @param {string} [options.secret = paypal_secret] - PayPal Secret.
+ *
+ * @returns {Promise<object>} Settings for paypal.
+ */
+const getOptions = async ({clientId, secret, api} = getConfig()) => {
+  const {access_token, token_type} = await getAuthToken(
+    {clientId, secret, api});
+  const Authorization = `${token_type} ${access_token}`;
```

Comment:
Could inline this in the headers object.

Reply:
good suggestion done here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/90f3edaba533e086ea7c70567f8e0f0c0e4c65ca

Thanks.